### DISCUSSION TITLE
Why alternatives failed?

### DISCUSSION THREAD

**Discussion by funny-falcon** [dstogov/ir#54] [Q&A] [answered]
[You've mentioned ](https://github.com/php/php-src/pull/12079#issuecomment-1718225662) you analyzed [MIR jit](https://github.com/vnmakarov/mir), libfirm and other alternatives.

Why did they fail for you? Especially interesting reasoning about MIR.

**Comment by dstogov** [comment, marked as answer]
> [You've mentioned ](https://github.com/php/php-src/pull/12079#issuecomment-1718225662) you analyzed [MIR jit](https://github.com/vnmakarov/mir), libfirm and other alternatives.
> Why did they fail for you?

You may find some comments in my presentation. See https://www.researchgate.net/publication/374470404_IR_JIT_Framework_a_base_for_the_next_generation_JIT_for_PHP (slide 10).

> Especially interesting reasoning about MIR.

MIR is an interesting project and I like its ideas, but at  time I analysed it I wasn't satisfied with generated code quality, absence of tooling (disassembler, gdb, perf), missing of x86 32-bit support. Finally MIR is actually slow (see https://github.com/vnmakarov/mir/issues/84)  

Do you use MIR?
